error due to insufficient alloc parameters and proc-macro derive panick

I'm using `rustc 1.44.0-nightly (b543afca9 2020-04-05)` on macOS Mojave 10.14.6.

After installing relevant dependencies with:
```
cargo clean
rustup toolchain install nightly
rustup default nightly

rustup target add wasm32-wasi
```

When I then try to run `cargo build --verbose` to run the "hosted kernel" I get the following error:

```
   Compiling linked_list_allocator v0.8.1
     Running `rustc --crate-name linked_list_allocator /Users/ls/.cargo/registry/src/github.com-1ecc6299db9ec823/linked_list_allocator-0.8.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on --cfg 'feature="alloc_ref"' --cfg 'feature="default"' --cfg 'feature="spinning_top"' --cfg 'feature="use_spin"' -C metadata=ac322e2358cfcf0a -C extra-filename=-ac322e2358cfcf0a --out-dir /Users/ls/code/src/ltfschoen/redshirt/target/debug/deps -L dependency=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps --extern spinning_top=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libspinning_top-5ef7c9fb980d2472.rmeta --cap-lints allow`
error[E0050]: method `alloc` has 2 parameters but the declaration in trait `core::alloc::AllocRef::alloc` has 3
   --> /Users/ls/.cargo/registry/src/github.com-1ecc6299db9ec823/linked_list_allocator-0.8.1/src/lib.rs:136:14
    |
136 |     fn alloc(&mut self, layout: Layout) -> Result<(NonNull<u8>, usize), AllocErr> {
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 parameters, found 2
    |
    = note: `alloc` from trait: `fn(&mut Self, core::alloc::Layout, core::alloc::AllocInit) -> core::result::Result<core::alloc::MemoryBlock, core::alloc::AllocErr>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0050`.
error: could not compile `linked_list_allocator`.

Caused by:
  process didn't exit successfully: `rustc --crate-name linked_list_allocator /Users/ls/.cargo/registry/src/github.com-1ecc6299db9ec823/linked_list_allocator-0.8.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on --cfg 'feature="alloc_ref"' --cfg 'feature="default"' --cfg 'feature="spinning_top"' --cfg 'feature="use_spin"' -C metadata=ac322e2358cfcf0a -C extra-filename=-ac322e2358cfcf0a --out-dir /Users/ls/code/src/ltfschoen/redshirt/target/debug/deps -L dependency=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps --extern spinning_top=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libspinning_top-5ef7c9fb980d2472.rmeta --cap-lints allow` (exit code: 1)
warning: build failed, waiting for other jobs to finish...
error: build failed
```

And if I run `cargo run --verbose` to run the "hosted kernel" I get the following error:

```
   Compiling redshirt-cli-kernel v0.1.0 (/Users/ls/code/src/ltfschoen/redshirt/kernel/cli)
     Running `rustc --crate-name redshirt_cli_kernel --edition=2018 kernel/cli/src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on -C metadata=6bcc27360ca0b8d4 -C extra-filename=-6bcc27360ca0b8d4 --out-dir /Users/ls/code/src/ltfschoen/redshirt/target/debug/deps -C incremental=/Users/ls/code/src/ltfschoen/redshirt/target/debug/incremental -L dependency=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps --extern async_std=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libasync_std-b5218eeeadc643ff.rlib --extern futures=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libfutures-32890a7a54457d0e.rlib --extern parity_scale_codec=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libparity_scale_codec-100caa81be8a32f1.rlib --extern redshirt_core=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_core-f53b8497aa46036f.rlib --extern redshirt_log_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_log_hosted-3b72cc9002d0d424.rlib --extern redshirt_random_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_random_hosted-c8cdc89d3719d40f.rlib --extern redshirt_syscalls=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_syscalls-4d4216f6e156dab6.rlib --extern redshirt_tcp_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_tcp_hosted-2096ee06237b8e10.rlib --extern redshirt_time_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_time_hosted-0dd44ee2516158fe.rlib --extern structopt=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libstructopt-74ba47ba47afdf73.rlib --extern wasi=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libwasi-01e260c710d1cced.rlib`
error: proc-macro derive panicked
  --> kernel/cli/src/main.rs:76:31
   |
76 |           .with_startup_process(build_wasm_module!(
   |  _______________________________^
77 | |             "../../../modules/p2p-loader",
78 | |             "modules-loader"
79 | |         ))
   | |_________^
   |
   = help: message: assertion failed: output.status.success()
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `proc_macro_call` in this scope
  --> kernel/cli/src/main.rs:76:31
   |
76 |           .with_startup_process(build_wasm_module!(
   |  _______________________________^
77 | |             "../../../modules/p2p-loader",
78 | |             "modules-loader"
79 | |         ))
   | |_________^
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

error: could not compile `redshirt-cli-kernel`.

Caused by:
  process didn't exit successfully: `rustc --crate-name redshirt_cli_kernel --edition=2018 kernel/cli/src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on -C metadata=6bcc27360ca0b8d4 -C extra-filename=-6bcc27360ca0b8d4 --out-dir /Users/ls/code/src/ltfschoen/redshirt/target/debug/deps -C incremental=/Users/ls/code/src/ltfschoen/redshirt/target/debug/incremental -L dependency=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps --extern async_std=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libasync_std-b5218eeeadc643ff.rlib --extern futures=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libfutures-32890a7a54457d0e.rlib --extern parity_scale_codec=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libparity_scale_codec-100caa81be8a32f1.rlib --extern redshirt_core=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_core-f53b8497aa46036f.rlib --extern redshirt_log_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_log_hosted-3b72cc9002d0d424.rlib --extern redshirt_random_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_random_hosted-c8cdc89d3719d40f.rlib --extern redshirt_syscalls=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_syscalls-4d4216f6e156dab6.rlib --extern redshirt_tcp_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_tcp_hosted-2096ee06237b8e10.rlib --extern redshirt_time_hosted=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libredshirt_time_hosted-0dd44ee2516158fe.rlib --extern structopt=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libstructopt-74ba47ba47afdf73.rlib --extern wasi=/Users/ls/code/src/ltfschoen/redshirt/target/debug/deps/libwasi-01e260c710d1cced.rlib` (exit code: 1)
```

--- a/README.md
+++ b/README.md
@@ -1,5 +1,5 @@
 The **redshirt** operating system is an experiment to build some kind of operating-system-like
-environment where executables are all in WASM and are loaded from an IPFS-like decentralized
+environment where executables are all in Wasm and are loaded from an IPFS-like decentralized
 network.
 
 I'm frequently telling people what my vision of an operating system would be. Now I've started
@@ -9,48 +9,46 @@ building it.
 
 # How to test
 
-There are two binaries available in this repository:
+**Important**: At the moment, most of the compilation requires a nightly version of Rust. See also https://github.com/tomaka/redshirt/issues/300.
 
-- The "hosted kernel" is a regular binary that executes WASM programs and uses the host operating
-  system.
-- The freestanding kernel is a multiboot2-compliant kernel that can be loaded with GRUB2 or any
-  compliant bootloader.
-
-For the hosted kernel:
+You also need to install the `wasm32-wasi` target, as the Wasm modules are compiled for Wasi.
 
 ```
-# You need the WASI target installed:
-rustup target add wasm32-wasi
+rustup toolchain install --target=wasm32-wasi nightly
+```
 
-# Then:
-cargo run
+There are two binaries available in this repository:
+
+- The "CLI kernel" is a regular binary that executes Wasi programs and leverages functionalities from the host operating system.
+- The freestanding kernel is a bare-metal kernel.
+
+For the CLI kernel:
+
+```
+# TODO: `--module-hash` must be passed the hash of the module to load,
+# but there is no modules-hosting platform at t the moment
+# See https://github.com/tomaka/redshirt/issues/333
+cargo +nightly run -- --module-hash=A
 ```
 
 For the freestanding kernel:
 
 ```
-rustup target add wasm32-wasi
 cd kernel/standalone-builder
-cargo run -- emulator-run --emulator qemu --target arm-rpi2
-```
-
-The freestanding kernel also supports x86_64:
-
-```
-cargo run -- emulator-run --emulator qemu --target x86_64-multiboot2
+cargo +nightly run -- emulator-run --emulator qemu --target x86_64-multiboot2
 ```
 
 # Repository structure
 
 Short overview of the structure of the repository:
 
-- `core` is a crate containing all the core infrastructure of interpreting WASM and inter-process
+- `core` is a crate containing all the core infrastructure of interpreting Wasm and inter-process
   communication. It is meant to become `#![no_std]`-compatible.
-- `interfaces` contains crates that provide definitions and helpers for WASM programs to use
+- `interfaces` contains crates that provide definitions and helpers for Wasm programs to use
   (examples: `tcp` for TCP/IP, `window` for windowing).
 - `kernel` contains the kernel binaries, plus crates that implement interfaces using the host's
   environment (e.g.: implements the `tcp` interface using Linux's or Window's TCP/IP).
-- `modules` contains WASM programs.
+- `modules` contains Wasm programs.
 
 # Contributing
 
@@ -67,10 +65,10 @@ to get in touch if you want to contribute anything non-trivial.
   or something similar.
 
 - If it ever becomes a real OS, everything would be run in ring 0. Isolation is guaranteed by the
-  WASM interpreter, and no hardware capability is required.
+  Wasm interpreter, and no hardware capability is required.
 
 - Programs are referred to by their hash, not by a file name. For example you don't tell the OS
-  "execute /usr/bin/foo". Instead you say "execute A45d9a21c3a7". The WASM binary, if it doesn't
+  "execute /usr/bin/foo". Instead you say "execute A45d9a21c3a7". The Wasm binary, if it doesn't
   exist locally, is fetched from a peer-to-peer network similar to IPFS.
 
 - There exists 3 core syscalls (send a message, send an answer, wait for a notification), and
@@ -86,7 +84,7 @@ to get in touch if you want to contribute anything non-trivial.
 - Low-level interfaces are handled by the kernel itself. On desktop, the kernel handles for example
   TCP/IP, UDP, file system, etc. by asking the host OS. On bare metal, the provided interfaces would
   be for example "interrupt handler manager", "PCI", etc. and the handler for the TCP/IP interface
-  would be a regular WASM process that communicates with the PCI, Ethernet, etc. interfaces.
+  would be a regular Wasm process that communicates with the PCI, Ethernet, etc. interfaces.
 
 - For security purposes, the user could choose to grant or deny to access to interface on a
   per-program basis, much like Android/browsers/etc. do. Don't want "sudoku" to access TCP/IP?
@@ -100,4 +98,4 @@ to get in touch if you want to contribute anything non-trivial.
   becomes incompatible. There's no version number.
 
 - The programs loader is itself just an interface handler. In other words, when the kernel wants to
-  start a program, it sends an IPC message to a process that then returns the WASM bytecode.
+  start a program, it sends an IPC message to a process that then returns the Wasm bytecode.

